CPDShell/Core/algorithms/knn_algorithm.py

```diff
@@ -0,0 +1,175 @@
+
+        :param metric: function for calculating distance between points in time series.
+        :param k: number of neighbours in graph relative to each point.
+        :param threshold: threshold that statistics should overcome to fix change point.
+        """
+        self.__k = k
+        self.__metric = metric
+        self.__threshold = threshold
+
+        self.__change_points: list[int] = []
+        self.__change_points_count = 0
+
+        self.__knngraph: knngraph.KNNGraph | None = None
+
+    def detect(self, window: Iterable[float | np.float64]) -> int:
+        """Finds change points in window.
+
+        :param window: part of global data for finding change points.
+        :return: the number of change points in the window.
+        """
+        self.__process_data(False, window)
+        return self.__change_points_count
+
+    def localize(self, window: Iterable[float | np.float64]) -> list[int]:
+        """Finds coordinates of change points (localizes them) in window.
+
+        :param window: part of global data for finding change points.
+        :return: list of window change points.
+        """
+        self.__process_data(window)
+        return self.__change_points.copy()
+
+    def __process_data(self, window: Iterable[float | np.float64]) -> None:
+        """
+        Processes a window of data to detect/localize all change points depending on working mode.
+
+        :param window: part of global data for change points analysis.
+        """
+        sample = deque(window)
+        sample_size = len(sample)
+        if sample_size == 0:
+            return
+
+        # Preparing.
+        self.__change_points: list[int] = []
+        self.__change_points_count = 0
+
+        # Building graph.
+        self.__knngraph = knngraph.KNNGraph(window, self.__metric, self.__k)
+        self.__knngraph.build()
+
+        # Examining each point.
+        # Boundaries are always change points.
+        first_point = int(len(window) / 4)
+        last_point = int(len(window) * 0.75)
+
+        for time in range(first_point, last_point):
+            statistics = self.__calculate_statistics_in_point(time, len(window))
+            # print(time, statistics)
+            if self.__check_change_point(statistics):
+                self.__change_points.append(time)
+                self.__change_points_count += 1
+
+    def __calculate_statistics_in_point(self, time: int, window_size: int) -> float:
+        """
+        Calculate the statistics of the KNN graph in specified point.
+
+        :param time: index of point in the given sample to calculate statistics relative to it.
+        :param window_size: size of sample to analyze.
+        """
+        assert self.__knngraph is not None, "Graph should not be None."
+
+        k = self.__k
+        n = window_size
+        n_1 = time
+        n_2 = n - time
+
+        if n <= k:
+            # Unable to analyze sample due to its size.
```

Comment:
What happens in this case? What means a negative number? Assert seems an overkill, because it;s a vaild case, but it'b be good to record this behabior in documentation. And how about `-np.inf`? It won't cross the threshold for sure.